### PR TITLE
Repeatable directives support

### PR DESCRIPTION
**Pull request by bbakerman**
This is a starting branch on repeatable directives.

https://github.com/graphql/graphql-spec/pull/472/files

It supersedes the other PRs however code from them has been incorporated (via copy paste not commits) so we are grateful for those contributions

We have

* grammar
* AST directive definition support
* AST directive container support
* some tests of that
* GraphqlXXX runtime support of getting repeatable directives
* Introspection support for show if a directive is repeatable
* Tests for schema building with repeatable directives

### CODE REVIEW

**Comment by bbakerman** [comment]
See #1916 and #1986

File: src/main/java/graphql/language/InlineFragment.java [hunk 1]
```diff
@@ -68,14 +67,6 @@ public TypeName getTypeCondition() {
...
     }
 
-    public Map<String, Directive> getDirectivesByName() {
-        return directivesByName(directives);
-    }
-
-    public Directive getDirective(String directiveName) {
-        return getDirectivesByName().get(directiveName);
-    }
-
```
Comment (bbakerman) [inline src/main/java/graphql/language/InlineFragment.java:78]: Should never had had this repeated code but rather inherit it

File: src/main/java/graphql/schema/GraphQLEnumValueDefinition.java [hunk 2]
```diff
@@ -119,16 +119,6 @@ public String getDeprecationReason() {
...
-    @Override
-    public Map<String, GraphQLDirective> getDirectivesByName() {
-        return DirectivesUtil.directivesByName(directives);
-    }
-
-    @Override
-    public GraphQLDirective getDirective(String directiveName) {
-        return getDirectivesByName().get(directiveName);
-    }
-
```
Comment (bbakerman) [inline src/main/java/graphql/schema/GraphQLEnumValueDefinition.java:131]: Should never had had this code

File: src/main/java/graphql/language/DirectivesContainer.java [hunk 3]
```diff
@@ -6,34 +6,67 @@
...
- * Represents a language node that can contain Directives.
+ * Represents a language node that can contain Directives.  Directives can be repeatable and (by default) non repeatable.
+ * <p>
+ * There are access methods here that get the two different types.
+ *
+ * @see graphql.language.DirectiveDefinition
+ * @see DirectiveDefinition#isRepeatable()
  */
 @PublicApi
 public interface DirectivesContainer<T extends DirectivesContainer> extends Node<T> {
```
Comment (bbakerman) [inline src/main/java/graphql/language/DirectivesContainer.java:23]: This method pattern is repeated in GraphqlDirectiveContainer as well - same names - same pattern - different types

File: src/main/java/graphql/language/DirectiveDefinition.java [hunk 4]
```diff
@@ -140,6 +153,7 @@ public DirectiveDefinition transform(Consumer<Builder> builderConsumer) {
         private SourceLocation sourceLocation;
         private List<Comment> comments = new ArrayList<>();
         private String name;
+        private boolean repeatable = false;
```
Comment (andimarek) [inline src/main/java/graphql/language/DirectiveDefinition.java:156]: it is false by default :-)

**Review by dugenkui03 — COMMENTED** [review]
There seem to be some mistack and performance problem, i left some comment for reference.

File: src/main/java/graphql/language/DirectiveDefinition.java (hunk 4 above)
Comment (bbakerman) [reply src/main/java/graphql/language/DirectiveDefinition.java:156]: just making it clear! I know java defaults but hey - a readder will see this and know as well

**Comment by dugenkui03** [comment]
https://github.com/graphql-java/graphql-java/blob/61a78fe152d2b510539c7c4e660baf3d0830ae72/src/main/java/graphql/introspection/IntrospectionQuery.java#L17

Need "isRepeatable" in `IntrospectionQuery` for introspection.

File: build.gradle [hunk 5]
```diff
@@ -75,6 +76,8 @@ dependencies {
 
     testCompile 'org.openjdk.jmh:jmh-core:1.21'
     testCompile 'org.openjdk.jmh:jmh-generator-annprocess:1.21'
+
+    testCompile 'is.tagomor.woothee:woothee-java:1.11.0'
```
Comment (bbakerman) [inline build.gradle:80]: This is some some rubbish left over testing - this is NOT needed

File: src/main/java/graphql/introspection/Introspection.java [hunk 6]
```diff
@@ -430,13 +430,17 @@ private static String print(Object value, GraphQLInputType type) {
...
             .field(newFieldDefinition()
                     .name("description")
                     .type(GraphQLString))
+            .field(newFieldDefinition()
+                    .name("isRepeatable")
+                    .type(nonNull(GraphQLBoolean)))
             .field(newFieldDefinition()
                     .name("locations")
-                    .type(list(nonNull(__DirectiveLocation))))
+                    .type(nonNull(list(__DirectiveLocation))))
```
Comment (andimarek) [inline src/main/java/graphql/introspection/Introspection.java:443]: what is this fix here? did we have this wrong before?

File: src/main/java/graphql/language/NodeUtil.java [hunk 7]
```diff
@@ -28,9 +30,26 @@ public static boolean isEqualTo(String thisStr, String thatStr) {
...
 
+    public static Map<String, List<Directive>> allDirectivesByName(List<Directive> directives) {
+        return FpKit.groupingBy(directives, Directive::getName);
+    }
+
+    public static Map<String, Directive> nonRepeatableDirectivesByName(List<Directive> directives) {
+        Map<String, List<Directive>> map = allDirectivesByName(directives);
+        List<Directive> singletonDirectives = map.entrySet().stream()
+                // only those that have 1 non repeated entry
+                .filter(e -> e.getKey() != null && e.getValue().size() == 1)
```
[the remark this answers is not in the document]
Comment (andimarek) [reply src/main/java/graphql/language/NodeUtil.java:40]: @bbakerman so this method return only the directives used once?

File: src/test/groovy/graphql/language/AstPrinterTest.groovy [hunk 8]
```diff
@@ -605,11 +601,41 @@ extend input Input @directive {
...
         when:
-        def result = astPrinter.printAst(interfaceType)
+        def result = AstPrinter.printAstCompact(interfaceType)
 
         then:
         result == "extend interface Resource implements Node & Extra {}"
 
     }
 
+    def "directive definitions can be printed"() {
```
Comment (andimarek) [inline src/test/groovy/graphql/language/AstPrinterTest.groovy:611]: we should have a test which actually prints multiple directives on a Node

**Review by andimarek — CHANGES_REQUESTED** [review]
Great work overall!
See my comments.

File: src/main/java/graphql/introspection/Introspection.java (hunk 6 above)
Comment (bbakerman) [reply src/main/java/graphql/introspection/Introspection.java:443]: let me check - this came from other code that I started this branch from - a PR from @dugenkui03  I think.

File: src/main/java/graphql/language/NodeUtil.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (bbakerman) [reply src/main/java/graphql/language/NodeUtil.java]: I have killed this method and its one usage

File: src/main/java/graphql/introspection/Introspection.java (hunk 6 above)
Comment (bbakerman) [reply src/main/java/graphql/introspection/Introspection.java:443]: The spec is not explicit

> `locations` returns a List of __DirectiveLocation representing the valid locations this directive may be placed.

So I will make it

```
.field(newFieldDefinition()
                    .name("locations")
                    .type(nonNull(list(nonNull(__DirectiveLocation)))))
```

Comment (andimarek) [reply src/main/java/graphql/introspection/Introspection.java]: The spec is explicit (see https://spec.graphql.org/draft/#sec-Schema-Introspection):
```
type __Directive {
  name: String!
  description: String
  locations: [__DirectiveLocation!]!
  args: [__InputValue!]!
  isRepeatable: Boolean!
}
```

**Review by andimarek — APPROVED** [review]

**Pull request merged**